### PR TITLE
Feature/use datetimeprovider issue 519

### LINKED ISSUE #519: [DateTime] Find and fix all instances of DateTime/DateTimeOffset being used instead of IDateTimeProvider
For example ` Stratis.Bitcoin.Features.Consensus.CoinViews.BackendPerformanceCounter` is one such class. There are many others.

### PR DESCRIPTION
**Pull request by Aprogiena**
Thought it was low hanging fruit. And it was nightmare :D 

Introduced DateTimeProvider almost everywhere. The remaining cases are couple of static classes and are +- legitimate, or not replaceable now. Also I tried to use DateTime with DateTime and DateTimeOffset with DateTimeOffset, not to mix them, although I believe DateTimeOffset should be completely avoided. I also replaced all `Now`s with `UtcNow`s as I believe non-UTC time should only be used on UI side, not server side or backend side. I haven't touched NBitcoin, however!

Minor polishing included as usual.

Fix #519

### CODE REVIEW

**Comment by dangershony** [comment]
You will have to merge it again please

**Comment by Aprogiena** [comment]
Yes, of course. Done that, but it needs a lot of testing and careful review. This was maybe too big task to make it in one go.

**Comment by dangershony** [comment]
yeah and you will continue to accumulate conflicts as this touches so many files, do you think you can easily break it to smaller PRs?

**Comment by Aprogiena** [comment]
not really, we should rather focus on cleaning the PRs we have before advancing further, things are waiting for ages and this will happen not only to this PR but to others, so i think the best solution is to just clean the queue

this PR is so big because a simple change in one component affects so many files, so it is not really easy to make this multiple small PRs

I will fix the conflicts here once PR 492 is merged

**Comment by Aprogiena** [comment]
ok, this should be ready again

**Comment by dangershony** [comment]
Ah crap more conflicts, please fix the conflicts it and I will now review this PR next before anything else.

**Comment by Aprogiena** [comment]
No worries, I've expected this to happen. This is nightmarish PR. So thanks.

File: Stratis.Bitcoin.Features.Consensus/ConsensusPerformanceCounter.cs [hunk 1]
```diff
@@ -189,9 +190,17 @@ public DateTime Taken
...
+
+        /// <summary>
+        /// Initializes an instance of the object.
+        /// </summary>
+        /// <param name="dateTimeProvider">Provider of date time functionality.</param>
+        public ConsensusPerformanceCounter(IDateTimeProvider dateTimeProvider)
 		{
-            this._Start = DateTime.UtcNow;
+            this.dateTimeProvider = dateTimeProvider;
+            this._Start = this.dateTimeProvider.GetUtcNow();
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Consensus/ConsensusPerformanceCounter.cs]: `this._Start` -> `this.Start`

Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Consensus/ConsensusPerformanceCounter.cs]: these are off topics, we could make many new PRs fixing code style, tabs and comments, but if the old code was bad, it is OK not to fix; it is not OK to introduce new bad

File: Stratis.Bitcoin.Features.Wallet/Wallet.cs [hunk 2]
```diff
@@ -227,8 +228,16 @@ public IEnumerable<UnspentOutputReference> GetAllSpendableTransactions(CoinType
...
     {
-        public AccountRoot()
+        /// <summary>Provider of date time functionality.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>
+        /// Initializes a new instance of the wallet.
+        /// </summary>
+        /// <param name="dateTimeProvider">Provider of date time functionality.</param>
+        public AccountRoot(IDateTimeProvider dateTimeProvider)
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/Wallet.cs]: In this case I would prefer to not have `IDateTimeProvider` inside `AccountRoot`.
Rather pass the creation date to the constructor.

File: Stratis.Bitcoin.Features.Wallet/WalletExtensions.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ public static class WalletExtensions
         /// <param name="chain">The chain.</param>        
         public static bool IsDownloaded(this ConcurrentChain chain)
         {
-            return chain.Tip.Header.BlockTime.ToUnixTimeSeconds() > (DateTimeOffset.Now.ToUnixTimeSeconds() - TimeSpan.FromHours(1).TotalSeconds);
+            return chain.Tip.Header.BlockTime.ToUnixTimeSeconds() > (DateTimeOffset.UtcNow.ToUnixTimeSeconds() - TimeSpan.FromHours(1).TotalSeconds);
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Wallet/WalletExtensions.cs]: Hmm not the best but I guess its an exception

File: Stratis.Bitcoin.Features.Wallet/Wallet.cs (hunk 2 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Wallet/Wallet.cs]: yes i agree, i've made such decision elsewhere, but not here, but i don't like it either :)

File: Stratis.Bitcoin.Features.Wallet/WalletExtensions.cs (hunk 3 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Wallet/WalletExtensions.cs]: one of 2 or 3 exceptions only, no problem to have it though

File: Stratis.Bitcoin/Base/TimeSyncBehavior.cs [hunk 4]
```diff
@@ -450,7 +450,7 @@ private void AttachedNode_MessageReceived(Node node, IncomingMessage message)
                     VersionPayload version = node.PeerVersion;
                     if (version != null)
                     {
-                        TimeSpan timeOffset = version.Timestamp - this.dateTimeProvider.GetUtcNow();
+                        TimeSpan timeOffset = version.Timestamp - this.dateTimeProvider.GetTimeOffset();
```
Comment (dangershony) [inline Stratis.Bitcoin/Base/TimeSyncBehavior.cs]: Was that a bug?

Comment (Aprogiena) [reply Stratis.Bitcoin/Base/TimeSyncBehavior.cs]: not really, but i was quite strict here in this PR, to really use the same types everywhere
but in general, i would rather not use DateTimeOffset anywhere, but well ... it is there, so let's be consistent

File: Stratis.Bitcoin.Features.Consensus/ConsensusPerformanceCounter.cs (hunk 1 above)
Comment (dangershony) [reply Stratis.Bitcoin.Features.Consensus/ConsensusPerformanceCounter.cs]: ok yeah ofcourse

**Pull request merged**